shared/runtime/pyexec.c

```diff
@@ -139,15 +147,46 @@ static int parse_compile_execute(const void *source, mp_parse_input_kind_t input
             mp_hal_stdout_tx_strn("\x04", 1);
         }
 
-        // check for SystemExit
-        if (mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t *)nlr.ret_val)->type), MP_OBJ_FROM_PTR(&mp_type_SystemExit))) {
-            // at the moment, the value of SystemExit is unused
+        #if MICROPY_ENABLE_VM_ABORT
+        if (nlr.ret_val == NULL) { // abort
             ret = PYEXEC_FORCED_EXIT;
-        } else {
+        } else
+        #endif
+        #if MICROPY_ENABLE_EXIT_CODE_HANDLING
+        if (mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t *)nlr.ret_val)->type), MP_OBJ_FROM_PTR(&mp_type_KeyboardInterrupt))) { // keyboard interrupt
             mp_obj_print_exception(&mp_plat_print, MP_OBJ_FROM_PTR(nlr.ret_val));
+            ret = PYEXEC_KEYBOARD_INTERRUPT;
+        } else
+        #endif
+        if (mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t *)nlr.ret_val)->type), MP_OBJ_FROM_PTR(&mp_type_SystemExit))) { // system exit
+            #if MICROPY_ENABLE_EXIT_CODE_HANDLING
+            mp_obj_t val = mp_obj_exception_get_value(MP_OBJ_FROM_PTR(nlr.ret_val));
+            if (val != mp_const_none) {
+                if (mp_obj_is_small_int(val)) {
+                    ret = MP_OBJ_SMALL_INT_VALUE(val);
+                } else {
+                    mp_obj_print(val, PRINT_STR);
+                    mp_print_str(&mp_plat_print, "\n");
+                    ret = PYEXEC_UNHANDLED_EXCEPTION;
+                }
+            } else {
+                ret = 0;
+            }
+            #else
+            ret = PYEXEC_FORCED_EXIT;
+            #endif
+        } else { // other exception
+            mp_obj_print_exception(&mp_plat_print, MP_OBJ_FROM_PTR(nlr.ret_val));
```

Comment:
To eliminate duplication of code for this line in the case of a `KeyboardInterrupt` I suggest:
```c
} else { // other exception
    mp_obj_print_exception(&mp_plat_print, MP_OBJ_FROM_PTR(nlr.ret_val));
    ret = PYEXEC_UNHANDLED_EXCEPTION;
    #if MICROPY_ENABLE_EXIT_CODE_HANDLING
    if (mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t *)nlr.ret_val)->type), MP_OBJ_FROM_PTR(&mp_type_KeyboardInterrupt))) {
        ret = PYEXEC_KEYBOARD_INTERRUPT;
    }
    #endif
}
```
And then define `PYEXEC_UNHANDLED_EXCEPTION` as 0 when `MICROPY_ENABLE_EXIT_CODE_HANDLING` is disabled.